### ISSUE TITLE
Compiler fails to remove jumps using jump table if every label in it is same

### ISSUE THREAD

**Issue by AngelicosPhosphoros** [rust-lang/rust#110797]
I tried this code:

```rust
pub enum K{
    A(Box<[i32]>),
    B(Box<[u8]>),
    C(Box<[String]>),
    D(Box<[u16]>),
}

pub fn get_len(arg: &K)->usize{
    match arg {
        K::A(ref lst)=>lst.len(),
        K::B(ref lst)=>lst.len(),
        K::C(ref lst)=>lst.len(),
        K::D(ref lst)=>lst.len(),
    }
}
```

I expected to see this happen:

Since layout of each enum variant is same, I expected to get machine code without any jumps, e.g.
```asm
example::get_len:
        mov     rax, qword ptr [rdi + 16]
        ret
```

Instead, this happened:
Compiler generates jump table, every branch of it points to the same code, and code jumps using it despite being it redundant.
```asm
example::get_len:
        mov     rax, qword ptr [rdi]
        lea     rcx, [rip + .LJTI0_0]
        movsxd  rax, dword ptr [rcx + 4*rax]
        add     rax, rcx
        jmp     rax
.LBB0_1:
        mov     rax, qword ptr [rdi + 16]
        ret
.LJTI0_0:                                     ; Note that each of it jumps to same label at the end.
        .long   .LBB0_1-.LJTI0_0
        .long   .LBB0_1-.LJTI0_0
        .long   .LBB0_1-.LJTI0_0
        .long   .LBB0_1-.LJTI0_0
```

### Meta

`rustc --version --verbose`:
```
rustc 1.68.0 (2c8cc3432 2023-03-06)
binary: rustc
commit-hash: 2c8cc343237b8f7d5a3c3703e3a87f2eb2c54a74
commit-date: 2023-03-06
host: x86_64-unknown-linux-gnu
release: 1.68.0
LLVM version: 15.0.6
Compiler returned: 0
```

Also, [godbolt link](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,selection:(endColumn:34,endLineNumber:12,positionColumn:34,positionLineNumber:12,selectionStartColumn:34,selectionStartLineNumber:12,startColumn:34,startLineNumber:12),source:'pub+enum+K%7B%0A++++A(Box%3C%5Bi32%5D%3E),%0A++++B(Box%3C%5Bu8%5D%3E),%0A++++C(Box%3C%5BString%5D%3E),%0A++++D(Box%3C%5Bu16%5D%3E),%0A%7D%0A%0Apub+fn+get_len(arg:+%26K)-%3Eusize%7B%0A++++match+arg+%7B%0A++++++++K::A(ref+lst)%3D%3Elst.len(),%0A++++++++K::B(ref+lst)%3D%3Elst.len(),%0A++++++++K::C(ref+lst)%3D%3Elst.len(),%0A++++++++K::D(ref+lst)%3D%3Elst.len(),%0A++++%7D%0A%7D%0A'),l:'5',n:'0',o:'Rust+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:r1680,deviceViewOpen:'1',filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,libs:!(),options:'-Copt-level%3D3',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+rustc+1.68.0+(Editor+%231)',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',m:100,n:'0',o:'',t:'0')),version:4).

**Comment by taiki-e** [comment]
This seems a regression introduced in Rust 1.65: https://godbolt.org/z/7vnT1z67E

@rustbot label +regression-from-stable-to-stable

**Comment by nikic** [comment]
It's another typed GEP issue.

**Comment by AngelicosPhosphoros** [comment]
https://github.com/llvm/llvm-project/issues/62570

**Comment by erikdesjardins** [comment]
Fixed in 1.73, presumably due to the LLVM 17 upgrade (#114048): https://godbolt.org/z/WTcKvcfh4

This specific example seems to be a backend optimization (the optimized IR is the same in 1.72 and 1.73); the general case is fixed in nightly by #121665: https://godbolt.org/z/cxnsWoWzx.

@rustbot label +E-needs-test
